### ISSUE TITLE
Questioning foundations: Do we need two delete/deleteAll methods in DeleteDataSource?

### ISSUE THREAD

**Issue by vilanovi** [mobilejazz/harmony-reference#8]
Current definition of the `DeleteDataSource` interface:
```
public protocol DeleteDataSource : DataSource {
    func delete(_ query: Query) -> Future<Void>
    func deleteAll(_ query: Query) -> Future<Void>
}
```

Both `delete` and `deleteAll` methods have the same exact signature and can be tricky to understand which one do you need to use/implement.

My question is simple? do we really need two methods? couldn't we just use a single one?

 ```
public protocol DeleteDataSource : DataSource {
    func delete(_ query: Query) -> Future<Void>
}
```

I'm just trying to open discussion to confirm this (having 2 methods) is really what we want and we are not just only extending without a real reason the `xxx`/`xxxAll` pattern we have in `GetDataSource` and `PutDataSource` (which are required as we need an array of elements in one case).

**Comment by pablosproject** [comment]
I will add that (based on REST RFC), all delete methods (same as PUT and POST) must return the value that has been modified, in this case deleted, by the operation. This will make harmony specification for backend and mobile to converge to the same standard.
Of course the result can be @DiscardableResult

**Comment by RubenVot** [comment]
I like the idea of @pablosproject I would change de Future to return the deleted objects.

**Comment by vilanovi** [comment]
@pablosproject @RubenVot the current implementation is returning `Void` because quite often is complicated (or impossible) to fetch the list of deleted objects, and, in order to simplify the interface, we just decided to return `void`. However, I agree you have a valid point in order to make it more compliant with how regular DELETE actions work, we should return the deleted objects.

However, please, let's focus on this thread on my open question and not diverge to side conversations (which are still valid, but if you think you really want to take action on them, better to open a new issue).

Thanks,

**Comment by RubenVot** [comment]
But the comments we did actually answer your question (sort of).

If you have a Void return I do agree with you @vilanovi: _It doesn't make sense to have delete and deleteAll._
It can be even confusing because what does it mean "delete" when there are several occurrences for the query? Delete first? Delete random? 
If you delete one occurrence randomly and you don't return it, how should the UI know which one has been deleted? This pattern might be forcing the UI to do a _getAll_ after _delete_ in order to know what had happened.

However, returning the deleted objects in the Future might solve this issue and can give some meaning to _get_ and _getAll_

**In summary:** if you think that returning the objects in the future is not plausible I do agree _get_ and _getAll_ should be simplified.  If returning the objects is plausible I would go that way.

**Comment by vilanovi** [comment]
Here you have a good valid point @RubenVot :).

The behavior of "delete" actions depends exclusively on the `Query` and its implementation. Therefore, I wouldn't expect "random deletions" as you mention, but rather targeted-specific deletions of objects as described on the Query itself. Any more complex behavior could be implemented by a GET + DELETE combination within an interactor.

In any case, I think it would be beneficial to ask @joselufo if he recalls the reason why we made DELETE methods return `Void` and not the deleted objects. Once we agree on this, we can then decide if we want to maintain one or two delete methods.

**Comment by joselufo** [comment]
Two points here:
- Why did we implement two methods for delete action?
The initial idea was to give more context of what you were doing, for example, if you want to delete more than one just use `deleteAll` function and if not `delete` function. 

I agree with you guys that maybe it doesn't give us any benefits and it only add confusion. So I would say that it's safe to remove the `deleteAll()` function and just use the `delete()` function to execute all the queries. 

- Why did we implement delete actions without a return type or value?
First, I disagree with @pablosproject the RFC standard doesn't say that you should add the entity as the result of the operation. It give you three options. Here you can find more information about it. https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.7/

We didn't add the value as the response because the RFC standard was open and we didn't find any case where the response value was useful to use, so to avoid confusions we just don't return the value as the response.

Adding the response value in the delete actions could be dangerous in some cases, for example, image that you delete an object and the `DeleteRepository` give you the deleted model, you could return the same input value as the database library, third-party service, etc... doesn't give you the deleted model. 

What happen is that you are going to return the same value with all the information and for example you could have the identifier set and maybe your business logic react to something if the identifier is available so you could end up with weird behaviours. 

I would say that adding the deleted value to the response of the `DeleteRepository` doesn't give any benefit and the RFC standard doesn't enforce us to do it.

**Comment by RubenVot** [comment]
Alea iacta est

**Comment by vilanovi** [comment]
I think that the best answer will be the one that makes our life easier without losing flexibility.

My reasoning is:

1) 90% of cases, our delete calls won't require to get back the list of deleted objects --> we can have our delete methods return Void.

2) 10% of cases, our domain will require to know the exact deleted objects --> we can achieve this by creating a custom interactors that compose get + delete actions. 

Therefore, I would keep Void as main return type and only If case 2) starts to happen very often, then we can decide to change this and add a returning type T/[T] in delete methods.  Until then, no need to take any action.

Regarding delete vs deleteAll, even if the symmetry of having xxx and xxxAll as on Get and Put produces me an orgasm (my OCD says hello to you), I would vote towards making our life easier and just have one single delete method.

Thoughts?

**Comment by RubenVot** [comment]
+1 !

**Comment by pablosproject** [comment]
I agree on everything you guys said, but I have a need here: as most of you know, is common practice to return the deleted entity on the DELETE method (at least is what I saw in 99% of the REST API interfaces I have worked with). So in the backend harmony implementation, I am in the need of having the returned entity.

We have 2 options:
1. Return the deleted entity (even if not contemplated by the RFC), but make it optional.
2. State and write in stone taht we are not doing this (also for backend API), and make everyone adapt to it.

**Comment by joselufo** [comment]
@pablosproject I think it's just better that you explain us why is really needed? maybe you could give us real examples where the delete entity is used so we can understand the necessity of it and maybe we can agree on having it. 

Taking into account Mobile, I think I never used the delete entity response from the API to do anything but this comment doesn't mean that it shouldn't be provided.

**Comment by pablosproject** [comment]
@joselufo It's not really needed. I saw it done in all places (and is also part of the REST philosophy), but that is flexible. I used in some frontend apps, when I delete something, then I present a toast with the name of the deleted resources (without the need of saving the original one somewhere).

I have no problem in changing it to not return anything, my point is to obtaining it as a team decision (and include the web team in this, too).

**Comment by vilanovi** [comment]
> my point is to obtain it as a team decision 

Yep! That's the goal here :).

Maybe @pablosproject everything goes down to what will be our most recurrent use case? Doesn't matter if others do it or not, what matters is what we do? Maybe more people from the web team could contribute and explain how they usually handle delete operations (deleting items from a list or similar). (And by web team, I also include you Paolo :P). 

> State and write in stone that we are not doing this (also for backend API), and make everyone adapt to it.

Anything is written in stone, but everything is acknowledged by the whole team. This is just a matter of agreeing in the most common scenario for everyone.

Furthermore, the mobile team could adapt to a new delete return type as @joselufo says, yes. However, note that it requires some effort and changes to update it in existing projects, and before deciding on this change (or fork the delete data source API between mobile and web) I think it would be nice to do a bit more of research on if this is really needed. "Out There" many people are doing many different things. What matters is what we need and how we want to do it.

**Comment by pablosproject** [comment]
To me it doesn't matter at the moment, I see no real advantages in it, so I am happy to keep the actual return type. Let's just enforce it in the harmony repository for web stuff.

**Comment by vilanovi** [comment]
Deal then. We can come back to this issue whenever you guys (or anyone) find significant issues on having no return type in delete methods.

By the way, do we have an agreement as well on having only one delete method? (and remove deleteAll?)

**Comment by joselufo** [comment]
I'll be happy to just have one method in the delete repository / datasource but I don't know why my feelings are that there is something that we are not taking into account and if we remove the deleteAll() function, it's going to punch us in our faces in the future.

**Comment by vilanovi** [comment]
Honestly, at this stage I would just deprecate this issue and if this raises up in the future, we can then decide again for delete vs deleteAll. Let's do nothing for now.

**Comment by franmontiel** [comment]
Even if no action is taken now, I think it makes sense to keep only the delete method and rely entirely on queries for deleting more than one object.

**Comment by joselufo** [comment]
@franmontiel ok that's makes sense. Let's try to rely only with the delete() function and the queries and let's see in the future what happen.

**Comment by franmontiel** [comment]
Let's try to move this forward or close it.

If we want to make a soft transition to a delete usage we can deprecate the deleteAll method but not remove it and if no significant drawbacks appear we can remove it. Also deleteAll implementation in this case should just redirect the call to the delete method.  What do you think?

**Comment by vilanovi** [comment]
Agreement:

- Move logic from `deleteAll` to `delete`
- Make `deleteAll` call `delete`
- Mark `deleteAll` deprecated
